doc/sphinx/proof-engine/ltac.rst

```diff
@@ -936,38 +1192,80 @@ We can perform pattern matching on goals using the following expression:
       with the :n:`match reverse goal with` variant.
 
-   .. tacv:: multimatch goal with {+| {+, @context_hyp} |- @cpattern => @ltac_expr } | _ => @ltac_expr end
+Examples:
 
-      Using :n:`multimatch` instead of :n:`match` will allow subsequent tactics
-      to backtrack into a right-hand side tactic which has backtracking points
-      left and trigger the selection of a new matching branch or combination of
-      hypotheses when all the backtracking points of the right-hand side have
-      been consumed.
+.. _match_goal_hyps_ex:
 
-      The syntax :n:`match [reverse] goal …` is, in fact, a shorthand for
-      :n:`once multimatch [reverse] goal …`.
+   .. example:: Matching hypotheses
 
-   .. tacv:: lazymatch goal with {+| {+, @context_hyp} |- @cpattern => @ltac_expr } | _ => @ltac_expr end
+      Hypotheses are matched from the last hypothesis (which is by default the newest
+      hypothesis) to the first until the :tacn:`apply` succeeds.
 
-      Using lazymatch instead of match will perform the same pattern matching
-      procedure but will commit to the first matching branch with the first
-      matching combination of hypotheses rather than trying a new matching if
-      the right-hand side fails. If the right-hand side of the selected branch
-      is a tactic with backtracking points, then subsequent failures cause
-      this tactic to backtrack.
+      .. coqtop:: reset all
+
+         Goal forall A B : Prop, A -> B -> (A->B).
+         intros.
+         match goal with
+         | H : _ |- _ => idtac "apply " H; apply H
+         end.
+
+.. _match_goal_hyps_rev_ex:
+
+   .. example:: Matching hypotheses with reverse
+
+      Hypotheses are matched from the first hypothesis to the last until the :tacn:`apply` succeeds.
+
+      .. coqtop:: reset all
+
+         Goal forall A B : Prop, A -> B -> (A->B).
+         intros.
+         match reverse goal with
+         | H : _ |- _ => idtac "apply " H; apply H
+         end.
+
+.. _match_goal_multiple_hyps_ex:
+
+   .. example:: Multiple ways to match hypotheses
+
+      Every possible match for the hypotheses is evaluated until the right-hand
+      side succeeds.  Note that `H1` and `H2` are never matched to the same hypothesis.
+      Observe that the number of permutations can grow as the factorial
+      of the number of hypotheses and hypothesis patterns.
+
+      .. coqtop:: reset all
+
+         Goal forall A B : Prop, A -> B -> (A->B).
+         intros A B H.
+         match goal with
+         | H1 : _, H2 : _ |- _ => idtac "match " H1 H2; fail
+         | _ => idtac
+         end.
+
+   .. todo need examples for:
+      match_context_rule ::= [ {*, @match_hyp } |- @match_pattern ] => @ltac_expr
+      match_hyp ::= | @name := {? [ @match_pattern ] : } @match_pattern
+
+.. todo PR The following items (up to numgoals) are part of "value_tactic".  I'd like to make
+   this a subsection and explain that they all return values.  How do I get a 5th-level section title?
```

Comment:
You just use a new symbol. Most of the time I use `^^^^^^^^^` after `~~~~~~~~~~~~`.

Reply:
That seems to work, however visually the difference between 4th and 5th levels is almost imperceptible (4th uses 20 point, 5th uses 18.4 point:

![image](https://user-images.githubusercontent.com/1253341/83336933-32c28d00-a26c-11ea-9eb4-ed78594d428b.png)

They don't show up in the table of contents, either, which I'd like:

![image](https://user-images.githubusercontent.com/1253341/83336961-5a195a00-a26c-11ea-95ad-0acc26fc8376.png)

Reply:
Indeed, this is what I was telling you earlier regarding the use of such a subsection level in the changelog. In the case of the changelog, the solution was to introduce an inline (local) TOC (cf. https://coq.github.io/doc/master/refman/changes.html#changes-in-8-12-beta1). In the case of the Ltac chapter though, I think the right solution is to have fewer subsections. The current situation with one heading per Ltac construct is an inheritance from before the Sphinx migration, but I don't think we should keep it as is. (Not saying it should be fixed in this PR though.)